### PR TITLE
Migrate spectral modifier functions to pvlib.spectrum

### PR DESCRIPTION
**Pull request by kandersolar**
- ~[ ] Closes #xxxx~
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

As suggested in https://github.com/pvlib/pvlib-python/pull/1062#issuecomment-696220404 and maybe elsewhere, this PR moves the functions for First Solar and SAPM spectral modifiers to the `pvlib.spectrum.mismatch` module.

~I also thought as long as we are moving things, it would be good to rename `sapm_spectral_loss` to `sapm_spectral_correction`.~ ~Currently, the two functions are renamed to just `spectrum.sapm` and `spectrum.first_solar`.~ Now renamed to `spectrum.spectral_factor_firstsolar` and `spectrum.spectral_factor_sapm`  Do others agree with this change?  Should `pvlib.pvsystem.PVSystem.sapm_spectral_loss` be renamed as well?

### CODE REVIEW

**Comment by adriesse** [comment]
Good move!  

Still thinking about the names and wondering what naming patterns they conform to and which patterns are currently gaining or losing popularity.  Subspecies (?) is often at the end.

File: docs/sphinx/source/reference/effects_on_pv_system_output/spectrum.rst [hunk 1]
```diff
@@ -10,3 +10,5 @@ Spectrum
    spectrum.get_example_spectral_response
    spectrum.get_am15g
    spectrum.calc_spectral_mismatch_field
+   spectrum.first_solar_spectral_correction
```
Comment (cwhanse) [inline docs/sphinx/source/reference/effects_on_pv_system_output/spectrum.rst:13]: Should this be `spectrum.mismatch.first_solar_spectral_correction` ?  Or does the import in `spectrum.__init__` handle the path?

File: docs/sphinx/source/reference/effects_on_pv_system_output/spectrum.rst [hunk 2]
```diff
@@ -10,3 +10,5 @@ Spectrum
...
+   spectrum.sapm_spectral_correction
```
Comment (cwhanse) [inline docs/sphinx/source/reference/effects_on_pv_system_output/spectrum.rst:14]: +1 on the rename.

File: docs/sphinx/source/reference/effects_on_pv_system_output/spectrum.rst (hunk 1 above)
Comment (kanderso-nrel) [reply docs/sphinx/source/reference/effects_on_pv_system_output/spectrum.rst:13]: > Or does the import in `spectrum.__init__` handle the path?

Yep.  I think sphinx's autodoc doesn't fetch docstrings by scanning `.py` files, but rather by importing the specified object and accessing its `__doc__` attribute.  So as long as the name is importable with the specified name it shouldn't matter where the original definition lives.  The choice here is whether we want `mismatch` to be included in the display text for the link.

**Comment by adriesse** [comment]
I'm not in favor of shortening the function names to just the author names.  I though we had some other discussion recently where that was discouraged actually.

**Comment by mikofski** [comment]
I had a similar thought recently, thanks for pulling this together!

**Comment by kanderso-nrel** [comment]
> I'm not in favor of shortening the function names to just the author names. I though we had some other discussion recently where that was discouraged actually.

My thought was that function names in other modules don't include the type of the model, e.g. it's not `iam.sapm_reflectance` or somesuch, so it makes sense to remove `spectral_modifier` from these names.  But that's a separate consideration from whether it's desirable to choose `sapm` and `first_solar` as the model names here.  Suggestions for alternatives are welcome :)

**Comment by cwhanse** [comment]
I don't know if this scales, but maybe we can reach out to authors and ask them what they'd like the function name to be. 

I agree with shortening the function names when the module conveys the function purpose, but, that creates a pit for new users to fall into:

```
from pvlib.iam import sapm
from pvlib.spectrum import sapm
```

etc.

Another thought to toss out: we could add a publication year to the function name to create space for e.g. another model from First Solar.

**Comment by adriesse** [comment]
I think the year may be added on occasion, but I don't think this is a global solution.  

There are few modules/namespaces as focused and conveniently short as ```iam```; because of these two characteristics it is not onerous to import iam and use iam.author().  We could try to create more of those, but again it would not be a global solution.

Spectrum does not contain only one type of function, so I think the functions need a suitable prefix.  The latest addition to this module got a nice long name: [calc_spectral_mismatch_field](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.spectrum.calc_spectral_mismatch_field.html#pvlib.spectrum.calc_spectral_mismatch_field) so if we take that as a starting point the new ones could be:

```calc_spectral_factor_sapm```
```calc_spectral_factor_firstsolar```

I somewhat hesitate to use *spectral_mismatch* in the names of the empirical models or approximations.

I'm also not sure about the ```calc_``` prefix because it doesn't seem to add much information, but I would go with the flow on that one.  Maybe we could introduce a prefix ```est_``` for functions that merely estimate things.  :)

**Comment by cwhanse** [comment]
```
spectral_factor_sapm
spectral_factor_firstsolar
```

These work for me.

> I somewhat hesitate to use spectral_mismatch in the names

Ditto.

**Comment by kanderso-nrel** [comment]
I have updated to `spectral_factor_sapm` and `spectral_factor_firstsolar`, but still grumble to myself that there is no `irradiance.transposition_perez` or `irradiance.decomposition_erbs` ;)

**Comment by adriesse** [comment]
I'll be anticipating some PR's with renames then!

File: pvlib/atmosphere.py [hunk 3]
```diff
@@ -336,175 +336,10 @@ def gueymard94_pw(temp_air, relative_humidity):
...
-    # Evaluate Spectral Shift
-    coeff = coefficients
-    ama = airmass_absolute
-    modifier = (
-        coeff[0] + coeff[1]*ama + coeff[2]*pw + coeff[3]*np.sqrt(ama) +
-        coeff[4]*np.sqrt(pw) + coeff[5]*ama/np.sqrt(pw))
-
-    return modifier
+first_solar_spectral_correction = deprecated(
+    since='0.9.6',
```
Comment (cwhanse) [inline pvlib/atmosphere.py:340]: Don't forget to update the `since` line if we're going to 0.10.0

**Review by cwhanse — APPROVED** [review]

**Comment by kandersolar** [comment]
Looking back through the discussion, it seems like the current state of the PR is acceptable to those who expressed opinions on the matter.  I'll go ahead and merge to prevent it from languishing again.

Codecov failure can be ignored.

**Pull request merged**